// react-native-permissions/lib/module/module.android.js
import { NativeModules, PermissionsAndroid as Core } from 'react-native';
import { RESULTS } from './constants';
import { uniq } from './utils';
const RNP = NativeModules.RNPermissions;

function grantedToStatus(granted) {
  return granted ? RESULTS.GRANTED : RESULTS.DENIED;
}

function coreStatusToStatus(status) {
  switch (status) {
    case 'granted':
      return RESULTS.GRANTED;

    case 'denied':
      return RESULTS.DENIED;

    case 'never_ask_again':
      return RESULTS.BLOCKED;

    default:
      return RESULTS.UNAVAILABLE;
  }
}

async function openSettings() {
  await RNP.openSettings();
}

async function check(permission) {
  if (!RNP.available.includes(permission)) {
    return RESULTS.UNAVAILABLE;
  }

  if (await RNP.isNonRequestable(permission)) {
    return RESULTS.BLOCKED;
  }

  const granted = await Core.check(permission);
  return grantedToStatus(granted);
}

async function request(permission, rationale) {
  if (!RNP.available.includes(permission)) {
    return RESULTS.UNAVAILABLE;
  }

  if (await RNP.isNonRequestable(permission)) {
    return RESULTS.BLOCKED;
  }

  const status = coreStatusToStatus((await Core.request(permission, rationale)));

  if (status === RESULTS.BLOCKED) {
    await RNP.setNonRequestable(permission);
  }

  return status;
}

async function splitByUsability(permissions) {
  const unusables = {};
  const usables = [];
  const blocklist = await RNP.getNonRequestables();

  for (let index = 0; index < permissions.length; index++) {
    const permission = permissions[index];

    if (blocklist.includes(permission)) {
      unusables[permission] = RESULTS.BLOCKED;
      continue;
    }

    if (!RNP.available.includes(permission)) {
      unusables[permission] = RESULTS.UNAVAILABLE;
      continue;
    }

    usables.push(permission);
  }

  return {
    unusables,
    usables
  };
}

function checkNotifications() {
  return RNP.checkNotifications();
}

async function checkMultiple(permissions) {
  const dedup = uniq(permissions);
  const {
    unusables: output,
    usables
  } = await splitByUsability(dedup);
  await Promise.all(usables.map(async permission => {
    const granted = await Core.check(permission);
    output[permission] = grantedToStatus(granted);
  }));
  return output;
}

async function requestMultiple(permissions) {
  const toSetAsNonRequestable = [];
  const dedup = uniq(permissions);
  const {
    unusables: output,
    usables
  } = await splitByUsability(dedup);
  const statuses = await Core.requestMultiple(usables);

  for (const permission in statuses) {
    if (statuses.hasOwnProperty(permission)) {
      const status = coreStatusToStatus(statuses[permission]);
      output[permission] = status;
      status === RESULTS.BLOCKED && toSetAsNonRequestable.push(permission);
    }
  }

  if (toSetAsNonRequestable.length > 0) {
    await RNP.setNonRequestables(toSetAsNonRequestable);
  }

  return output;
}

export const module = {
  openSettings,
  check,
  request,
  checkNotifications,
  requestNotifications: checkNotifications,
  checkMultiple,
  requestMultiple
};
